File: src/backend/app/schemas/molecule.py
"""
Pydantic schema models for molecule data validation, serialization, and API responses.

This module defines the data models used for molecule-related operations in the 
Molecular Data Management and CRO Integration Platform, with appropriate validation 
rules, type annotations, and documentation.
"""

from pydantic import BaseModel, Field, validator
from typing import Optional, List, Dict, Any
from uuid import UUID
from datetime import datetime

from ..models.molecule import Molecule
from .molecule_property import MoleculePropertyBase, MoleculePropertyCreate, MoleculePropertyResponse
from .user import UserResponse
from ..constants import MOLECULE_FLAG_STATUS
from ..utils.validation_utils import validate_smiles, validate_molecular_property, validate_molecule_flag_status


class MoleculeBase(BaseModel):
    """Base Pydantic model for molecule data with common fields"""
    smiles: str = Field(..., 
                        description="SMILES representation of the molecular structure",
                        example="CCO")
    flag_status: Optional[str] = Field(None, 
                                      description="Optional flag for priority review (e.g., 'HIGH', 'MEDIUM', 'LOW')",
                                      example="HIGH")
    properties: Optional[List[MoleculePropertyCreate]] = Field(default_factory=list,
                                                            description="List of molecular properties")

    @validator('smiles', pre=True)
    def validate_smiles(cls, v):
        """Validates that a SMILES string is properly formatted and represents a valid molecular structure"""
        if not v or not isinstance(v, str):
            raise ValueError("SMILES string cannot be empty")
        
        if not validate_smiles(v):
            raise ValueError(f"Invalid SMILES structure: {v}")
        
        return v

    @validator('properties', pre=True)
    def validate_properties(cls, v):
        """Validates that molecule properties are properly formatted and contain valid values"""
        if v is None:
            return []
        
        if not isinstance(v, list):
            raise ValueError("Properties must be a list")
        
        for prop in v:
            if not isinstance(prop, dict):
                raise ValueError("Each property must be a dictionary")
            
            if 'property_name' not in prop:
                raise ValueError("Each property must have a 'property_name' field")
            
            # Validate property name
            if not validate_molecular_property(prop['property_name'], prop.get('property_value')):
                raise ValueError(f"Invalid property: {prop['property_name']} with value {prop.get('property_value')}")
        
        return v

    @validator('flag_status', pre=True)
    def validate_flag_status(cls, v):
        """Validates that a flag status value is one of the allowed values"""
        if v is None:
            return None
        
        if not validate_molecule_flag_status(v):
            valid_statuses = list(MOLECULE_FLAG_STATUS.keys()) if MOLECULE_FLAG_STATUS else []
            raise ValueError(f"Invalid flag status: {v}. Must be one of {valid_statuses}")
        
        return v


class MoleculeCreate(MoleculeBase):
    """Pydantic model for creating a new molecule"""
    created_by: Optional[UUID] = Field(None, 
                                      description="ID of the user creating the molecule")


class MoleculeUpdate(BaseModel):
    """Pydantic model for updating an existing molecule"""
    flag_status: Optional[str] = Field(None, 
                                      description="Optional flag for priority review (e.g., 'HIGH', 'MEDIUM', 'LOW')",
                                      example="HIGH")
    properties: Optional[List[MoleculePropertyCreate]] = Field(None,
                                                           description="List of molecular properties to update")


class MoleculeInDBBase(MoleculeBase):
    """Base Pydantic model for molecule data as stored in the database"""
    id: UUID = Field(..., description="Unique identifier for the molecule")
    created_by: UUID = Field(..., description="ID of the user who created this molecule")
    created_at: datetime = Field(..., description="Timestamp when the molecule was created")
    updated_at: Optional[datetime] = Field(None, description="Timestamp when the molecule was last updated")

    class Config:
        from_attributes = True


class MoleculeResponse(MoleculeInDBBase):
    """Pydantic model for molecule data in API responses"""
    properties: List[MoleculePropertyResponse] = Field(default_factory=list,
                                                   description="List of molecular properties")
    creator: Optional[UserResponse] = Field(None, description="User who created this molecule")

    class Config:
        from_attributes = True


class MoleculeDetailResponse(MoleculeResponse):
    """Pydantic model for detailed molecule data in API responses"""
    image_url: Optional[str] = Field(None, 
                                    description="URL to a rendered image of the molecular structure",
                                    example="/api/v1/molecules/1/image")

    class Config:
        from_attributes = True


class MoleculeListResponse(BaseModel):
    """Pydantic model for paginated list of molecules in API responses"""
    items: List[MoleculeResponse] = Field(..., description="List of molecules")
    total: int = Field(..., description="Total number of molecules matching the criteria")
    page: int = Field(..., description="Current page number")
    size: int = Field(..., description="Number of items per page")


class MoleculeFilter(BaseModel):
    """Pydantic model for filtering molecules by various criteria"""
    smiles: Optional[str] = Field(None, description="Filter by SMILES string (partial match)")
    created_by: Optional[UUID] = Field(None, description="Filter by creator user ID")
    flag_status: Optional[str] = Field(None, description="Filter by flag status")
    properties: Optional[List[Dict]] = Field(None, description="Filter by property values")
    created_after: Optional[datetime] = Field(None, description="Filter by creation date (after)")
    created_before: Optional[datetime] = Field(None, description="Filter by creation date (before)")
    library_ids: Optional[List[UUID]] = Field(None, description="Filter by library membership")
    experiment_ids: Optional[List[UUID]] = Field(None, description="Filter by experiment membership")
    sort_by: Optional[str] = Field(None, description="Field to sort by")
    sort_desc: Optional[bool] = Field(False, description="Sort in descending order if true")

File: src/web/src/types/molecule.ts
import { UUID } from 'uuid';

/**
 * Enum for molecule flag status values
 * Used for prioritizing and categorizing molecules
 */
export enum FlagStatus {
  IMPORTANT = 'important',
  FAVORITE = 'favorite',
  REVIEW = 'review',
  ARCHIVED = 'archived'
}

/**
 * Enum for property filter operators
 * Used in the filtering system to allow complex property-based queries
 */
export enum FilterOperator {
  EQUALS = 'equals',
  NOT_EQUALS = 'not_equals',
  GREATER_THAN = 'greater_than',
  LESS_THAN = 'less_than',
  BETWEEN = 'between',
  CONTAINS = 'contains'
}

/**
 * Interface for molecular property data
 * Represents individual properties of a molecule with their values and metadata
 */
export interface MoleculeProperty {
  id: UUID;
  molecule_id: UUID;
  property_name: string;
  property_value: number | string | boolean;
  property_unit?: string;
  is_calculated: boolean;
}

/**
 * Interface for molecule data with all properties
 * Core data structure representing a chemical molecule in the system
 */
export interface Molecule {
  id: UUID;
  smiles: string;
  flag_status?: FlagStatus;
  created_by: UUID;
  created_at: string;
  updated_at?: string;
  properties: MoleculeProperty[];
  creator?: {
    id: UUID;
    email: string;
    username?: string;
  };
  image_url?: string;
}

/**
 * Interface for creating a new molecule
 * Contains the minimum required fields to create a molecule
 */
export interface MoleculeCreate {
  smiles: string;
  flag_status?: FlagStatus;
  properties?: Partial<MoleculeProperty>[];
}

/**
 * Interface for updating an existing molecule
 * Contains fields that can be modified after creation
 */
export interface MoleculeUpdate {
  flag_status?: FlagStatus;
  properties?: Partial<MoleculeProperty>[];
}

/**
 * Interface for filtering molecules by property values
 * Enables complex filtering based on molecular properties
 */
export interface PropertyFilter {
  name: string;
  value?: string | number | boolean;
  min_value?: number;
  max_value?: number;
  operator: FilterOperator;
}

/**
 * Interface for filtering molecules by various criteria
 * Comprehensive filtering options for molecule searches
 */
export interface MoleculeFilter {
  smiles_pattern?: string;
  flag_status?: FlagStatus;
  created_by?: UUID;
  property_filters?: PropertyFilter[];
  created_after?: string;
  created_before?: string;
  library_ids?: UUID[];
  experiment_ids?: UUID[];
  sort_by?: string;
  sort_desc?: boolean;
}

/**
 * Interface for paginated list of molecules in API responses
 * Used for returning large sets of molecule data with pagination
 */
export interface MoleculeListResponse {
  items: Molecule[];
  total: number;
  page: number;
  size: number;
}

/**
 * Interface for detailed molecule data with related entities
 * Provides comprehensive information about a molecule and its relationships
 */
export interface MoleculeDetailResponse {
  molecule: Molecule;
  libraries: {
    id: UUID;
    name: string;
  }[];
  experiments: {
    id: UUID;
    name: string;
    status: string;
    date?: string;
  }[];
}

/**
 * Interface for similarity search parameters
 * Used for finding molecules with similar structures
 */
export interface MoleculeSimilaritySearchParams {
  smiles: string;
  threshold?: number;
  limit?: number;
}

/**
 * Interface for substructure search parameters
 * Used for finding molecules containing a specific substructure
 */
export interface MoleculeSubstructureSearchParams {
  smarts: string;
  limit?: number;
}

/**
 * Interface for molecule state in Redux store
 * Represents the complete state of the molecule module in the application
 */
export interface MoleculeState {
  molecules: Molecule[];
  filter: MoleculeFilter;
  selectedMolecules: UUID[];
  loading: boolean;
  error: string | null;
  totalMolecules: number;
  currentPage: number;
  pageSize: number;
}